crates/sqlexec/src/background_jobs.rs

```diff
@@ -0,0 +1,326 @@
+use std::{
+    collections::HashMap,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc, Mutex,
+    },
+    time::Duration,
+};
+
+use async_trait::async_trait;
+use tokio::{
+    sync::{mpsc, oneshot},
+    task::JoinHandle,
+    time::Instant,
+};
+
+use crate::errors::{ExecError, Result};
+
+/// Collection of background jobs.
+pub struct JobRunner {
+    send: mpsc::UnboundedSender<ChannelMessage>,
+    listen: Arc<Mutex<Option<JoinHandle<()>>>>,
+}
+
+struct JobHandle {
+    join: JoinHandle<()>,
+    sleep: JoinHandle<()>,
+}
+
+enum ChannelMessage {
+    NewJob(BackgroundJob),
+    JobComplete(String, Result<()>),
+    HealthCheck,
+    Close,
+}
+
+impl JobRunner {
+    /// Create a new collection of background jobs.
+    pub fn new() -> Self {
+        let (send, recv) = mpsc::unbounded_channel();
+        // Spawn the listener for receiving messages.
+        let send_clone = send.clone();
+        let listen = tokio::spawn(async move { Self::listen(send_clone, recv).await });
+        let listen = Arc::new(Mutex::new(Some(listen)));
+        Self { send, listen }
+    }
+
+    /// Listens on the channel for new messages.
+    async fn listen(
+        send: mpsc::UnboundedSender<ChannelMessage>,
+        mut recv: mpsc::UnboundedReceiver<ChannelMessage>,
+    ) {
+        // Sends a "health check" mesage to the channel every two minutes.
+        let send_clone = send.clone();
+        tokio::spawn(async move {
+            let send = send_clone;
+            let mut health_check = tokio::time::interval(Duration::from_secs(2 * 60));
+            loop {
+                let _ = health_check.tick().await;
+                send.send(ChannelMessage::HealthCheck).unwrap();
+            }
+        });
```

Comment:
A health check exists in case a job panics, and we aren't able to add a new job. This problem automatically disappears if we go with the abovementioned comment by maintaining a `running_jobs` map. But that comes with different properties.